src/deno_dir.rs

```diff
@@ -189,27 +248,59 @@ impl DenoDir {
         base.join(module_specifier)?
       };
 
-    let mut p = j
-      .to_file_path()
-      .unwrap()
-      .into_os_string()
-      .into_string()
-      .unwrap();
+    match j.scheme() {
+      "file" => {
+        let mut p = j
+          .to_file_path()
+          .unwrap()
+          .into_os_string()
+          .into_string()
+          .unwrap();
+
+        if cfg!(target_os = "windows") {
+          // On windows, replace backward slashes to forward slashes.
+          // TODO(piscisaureus): This may not me be right, I just did it to make
+          // the tests pass.
+          p = p.replace("\\", "/");
+        }
 
-    if cfg!(target_os = "windows") {
-      // On windows, replace backward slashes to forward slashes.
-      // TODO(piscisaureus): This may not me be right, I just did it to make
-      // the tests pass.
-      p = p.replace("\\", "/");
+        module_name = p.to_string();
+        filename = p.to_string();
+      }
+      _ => {
+        module_name = module_specifier.to_string();
+        filename = get_cache_filename(self.deps.as_path(), j)
+          .to_str()
+          .unwrap()
+          .to_string();
+      }
     }
 
-    let module_name = p.to_string();
-    let filename = p.to_string();
-
+    debug!("module_name: {}, filename: {}", module_name, filename);
     Ok((module_name, filename))
   }
 }
 
+fn get_cache_filename(basedir: &Path, url: Url) -> PathBuf {
+  let mut out = basedir.to_path_buf();
+  out.push(url.host_str().unwrap());
+  for path_seg in url.path_segments().unwrap() {
```

Comment:
I have a hunch that we're gonna find a security hole or two in this code.
Maybe it'd be better to use `sha(path)` as the cache filename.